### PR TITLE
E2E:Testing FHIR Rest Api End Point.

### PR DESCRIPTION
**Pull request by sherrif10**
## Issue I worked on

see https://issues.openmrs.org/browse/FM2-    cc @ibacher . this is something like starting point before i dive into patient resource.

## Checklist: I completed these to help reviewers :)

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: fhir-server/resources/test.properties [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+# the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+#
+# Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+# graphic logo is a trademark of OpenMRS Inc.
+#
+
+fhir.resource.serverBase=http://localhost:8080/openmrs/ws/fhir2/R4
```
Comment (ibacher) [inline fhir-server/resources/test.properties:11]: Other than this one property, I'm not sure I see the need for the rest of this file?

File: fhir-server/pom.xml [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		<artifactId>fhir2</artifactId>
+		<groupId>org.openmrs.module</groupId>
+		<version>1.4.0-SNAPSHOT</version>
+	</parent>
+	<modelVersion>4.0.0</modelVersion>
+
+	<artifactId>fhir-server</artifactId>
+	<name>FHIR2-server Rest Client </name>
+
+	<dependencies>
```
Comment (ibacher) [inline fhir-server/pom.xml:13]: For this module we should have absolutely no dependencies on other parts of the FHIR2 module except maybe the test-data if you want to reuse those definitions. We probably don't need the dependency on OpenMRS either. We should have an explicit dependency on the hapi-fhir-client project.

File: fhir-server/Patient-client/validation/rest/HttClient.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.HttpServerErrorException;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+package org.openmrs.module.fhir2.fhir-server.validation.rest;
+
+public class HttClient {
```
Comment (ibacher) [inline fhir-server/Patient-client/validation/rest/HttClient.java:22]:
```suggestion
public class HttpClient {
```

File: fhir-server/Patient-client/validation/rest/ CustomRestTemplateCustomizer.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.fhir2.fhir-server.validation.rest;
+
+import org.springframework.boot.web.client.RestTemplateCustomizer;
+import org.springframework.web.client.RestTemplate;
+
+public interface  CustomRestTemplateCustomizer implements RestTemplateCustomizer {
```
Comment (ibacher) [inline fhir-server/Patient-client/validation/rest/ CustomRestTemplateCustomizer.java:15]:
```suggestion
public interface CustomRestTemplateCustomizer implements RestTemplateCustomizer {
```

File: fhir-server/Patient-client/r4/PatientRestClient.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.fhir2.Patient-client.r4;
+
+public class PatientRestClient {
```
Comment (ibacher) [inline fhir-server/Patient-client/r4/PatientRestClient.java:12]: This... doesn't look like valid Java code. I.e., most of it should live in some kind of function right?

File: omod/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      MODULE_WEB_ADMIN: 'false'
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:8080/openmrs/"]
+      timeout: 20s
+    volumes:
+      - openmrs-referenceapplication-data:/usr/local/tomcat/.OpenMRS
+      - ./activemq-data:/usr/local/tomcat/.OpenMRS/activemq-data
+
+volumes:
+  openmrs-referenceapplication-data:
```
Comment (ibacher) [inline omod/docker-compose.yml:40]: We really don't need volumes for non-persistent images.

```suggestion

```

**Review by ibacher — COMMENTED** [review]
I'm a little confused on the overall direction here. You have both the HAPI REST client and a bunch of Spring RestTemplate stuff. We really don't need both of those, afaict.

**Comment by sherrif10** [comment]
Thanks @ibacher  .Am addressing your comments.

File: fhir-server/Patient-client/r4/PatientRestClient.java (hunk 5 above)
Comment (sherrif10) [reply fhir-server/Patient-client/r4/PatientRestClient.java]: Right this for testing purpose.

File: fhir-server/pom.xml [hunk 7]
```diff
@@ -0,0 +1,116 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>fhir2</artifactId>
+        <groupId>org.openmrs.module</groupId>
+        <version>1.4.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>fhir-server</artifactId>
```
Comment (ibacher) [inline fhir-server/pom.xml:10]: We should probably have a better name for this, e.g., `full-stack-integration-tests` or something?

File: fhir-server/pom.xml [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    <name>FHIR2 Testing Rest EndPoints </name>
```
Comment (ibacher) [inline fhir-server/pom.xml:11]: Along the same spirit: `FHIR2 Full-stack Integration Tests`

File: fhir-server/pom.xml [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>com.fasterxml.jackson.core</groupId>
+                    <artifactId>jackson-databind</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>${project.parent.groupId}</groupId>
```
Comment (ibacher) [inline fhir-server/pom.xml:31]: Hopefully we don't actually need this at all?

File: fhir-server/Patient-client/r4/fhirStarterRestfulServer.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletResponse;
+import java.util.Arrays;
+import java.util.TimeZone;
+
+import org.openmrs.Patient;
+import org.springframework.context.ApplicationContext;
+
+public class fhirStarterRestfulServer extends RestfulServer {
```
Comment (ibacher) [inline fhir-server/Patient-client/r4/fhirStarterRestfulServer.java:35]: We really shouldn't need a new RestfulServer... ideally we'd use the existing one, right?

Comment (sherrif10) [reply fhir-server/Patient-client/r4/fhirStarterRestfulServer.java:35]: You right, i have just rectified that there is an already existing restful server.

**Comment by sherrif10** [comment]
@ibacher Have added a generic fhir client class that will handle transactions which will be triggered on server. Let me know if that makes sense.

**Comment by sherrif10** [comment]
Closing it, openinng a new pull request, this pull request looks to have much not neccessary.

**Pull request closed without merging**